File: pkg/ingestor/globalsort/codec.go
// Copyright 2023 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package globalsort

import (
	"encoding/binary"
)

// RangeProperty stores properties of a range:
// - Key: the start key of the range.
// - Offset: the start offset of the range in the file.
// - Size: the size of the range.
// - Keys: the number of keys in the range.
type RangeProperty struct {
	FirstKey []byte
	LastKey  []byte
	Offset   uint64
	Size     uint64
	Keys     uint64
}

// decodeMultiProps is only used for test.
func decodeMultiProps(data []byte) []*RangeProperty {
	var ret []*RangeProperty
	for len(data) > 0 {
		propLen := int(binary.BigEndian.Uint32(data))
		propBytes := data[4 : 4+propLen]
		rp := decodeProp(propBytes)
		ret = append(ret, rp)
		data = data[4+propLen:]
	}
	return ret
}

func decodeProp(data []byte) *RangeProperty {
	rp := &RangeProperty{}
	n := 0
	keyLen := int(binary.BigEndian.Uint32(data[n : n+4]))
	n += 4
	rp.FirstKey = data[n : n+keyLen]
	n += keyLen
	keyLen = int(binary.BigEndian.Uint32(data[n : n+4]))
	n += 4
	rp.LastKey = data[n : n+keyLen]
	n += keyLen
	rp.Size = binary.BigEndian.Uint64(data[n : n+8])
	n += 8
	rp.Keys = binary.BigEndian.Uint64(data[n : n+8])
	n += 8
	rp.Offset = binary.BigEndian.Uint64(data[n : n+8])
	return rp
}

// keyLen * 2 + p.size + p.keys + p.offset
const propertyLengthExceptKeys = 4*2 + 8 + 8 + 8

func encodeMultiProps(buf []byte, props []*RangeProperty) []byte {
	var propLen [4]byte
	for _, p := range props {
		binary.BigEndian.PutUint32(propLen[:],
			uint32(propertyLengthExceptKeys+len(p.FirstKey)+len(p.LastKey)))
		buf = append(buf, propLen[:4]...)
		buf = encodeProp(buf, p)
	}
	return buf
}

func encodeProp(buf []byte, r *RangeProperty) []byte {
	var b [8]byte
	binary.BigEndian.PutUint32(b[:], uint32(len(r.FirstKey)))
	buf = append(buf, b[:4]...)
	buf = append(buf, r.FirstKey...)
	binary.BigEndian.PutUint32(b[:], uint32(len(r.LastKey)))
	buf = append(buf, b[:4]...)
	buf = append(buf, r.LastKey...)
	binary.BigEndian.PutUint64(b[:], r.Size)
	buf = append(buf, b[:]...)
	binary.BigEndian.PutUint64(b[:], r.Keys)
	buf = append(buf, b[:]...)
	binary.BigEndian.PutUint64(b[:], r.Offset)
	buf = append(buf, b[:]...)
	return buf
}

File: rust/crates/tidb-codec/benches/codec.rs
// Copyright 2026 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Executable equivalents of every benchmark in `pkg/util/codec/bench_test.go`.

use std::hint::black_box;
use std::time::Instant;

use tidb_codec::{
    decode, decode_decimal, decode_one_typed, encode_bytes, encode_decimal_fixed, encode_int,
    encode_value, BYTES_FLAG,
};
use tidb_datatype::{Datum, Decimal, FieldType, FieldTypeCode};

const ITERATIONS: usize = 10_000;
const VALUE_COUNT: usize = 100;

fn measure(name: &str, mut operation: impl FnMut()) {
    let started = Instant::now();
    for _ in 0..ITERATIONS {
        operation();
    }
    println!("{name}: {:?}", started.elapsed());
}

fn compose_encoded_data() -> Vec<u8> {
    let values = (0..VALUE_COUNT)
        .map(|value| Datum::Int(value as i64))
        .collect::<Vec<_>>();
    encode_value(&values).unwrap()
}

fn benchmark_decode_with_size(encoded: &[u8]) {
    black_box(decode(black_box(encoded), VALUE_COUNT).unwrap());
}

fn benchmark_decode_with_out_size(encoded: &[u8]) {
    black_box(decode(black_box(encoded), 1).unwrap());
}

fn benchmark_encode_int_with_size() {
    let mut output = Vec::with_capacity(8);
    encode_int(&mut output, 10);
    black_box(output);
}

fn benchmark_encode_int_with_out_size() {
    let mut output = Vec::new();
    encode_int(&mut output, 10);
    black_box(output);
}

fn benchmark_decode_decimal(encoded: &[u8]) {
    black_box(decode_decimal(black_box(encoded)).unwrap());
}

fn benchmark_decode_one_to_chunk(encoded: &[u8], field_type: &FieldType) {
    black_box(decode_one_typed(black_box(encoded), field_type).unwrap());
}

fn main() {
    let encoded = compose_encoded_data();
    measure("BenchmarkDecodeWithSize", || {
        benchmark_decode_with_size(&encoded);
    });
    measure("BenchmarkDecodeWithOutSize", || {
        benchmark_decode_with_out_size(&encoded);
    });
    measure("BenchmarkEncodeIntWithSize", benchmark_encode_int_with_size);
    measure(
        "BenchmarkEncodeIntWithOutSize",
        benchmark_encode_int_with_out_size,
    );

    let decimal = Decimal::from_signed_literal("1211.1211113");
    let mut decimal_bytes = Vec::new();
    encode_decimal_fixed(&mut decimal_bytes, &decimal, 0, 0).unwrap();
    measure("BenchmarkDecodeDecimal", || {
        benchmark_decode_decimal(&decimal_bytes);
    });

    let mut raw = vec![BYTES_FLAG];
    encode_bytes(&mut raw, b"a");
    let integer = FieldType::new(FieldTypeCode::LongLong);
    measure("BenchmarkDecodeOneToChunk", || {
        benchmark_decode_one_to_chunk(&raw, &integer);
    });
}
